pkgs/native_assets_cli/lib/src/model/build_output.dart

```diff
@@ -2,143 +2,176 @@
-      assets: Asset.listFromYamlList(as<YamlList>(yamlMap[_assetsKey])),
+      assets: assets,
       dependencies:
           Dependencies.fromYaml(as<YamlList?>(yamlMap[_dependenciesKey])),
       metadata: Metadata.fromYaml(as<YamlMap?>(yamlMap[_metadataKey])),
     );
   }
 
-  Map<String, Object> toYaml() => {
+  Map<String, Object> toYaml(Version version) => {
         _timestampKey: timestamp.toString(),
-        _assetsKey: assets.toYaml(),
-        _dependenciesKey: dependencies.toYaml(),
-        _metadataKey: metadata.toYaml(),
+        _assetsKey: [
+          for (final asset in _assets) asset.toYaml(version),
+        ],
+        if (_dependencies.dependencies.isNotEmpty)
+          _dependenciesKey: _dependencies.toYaml(),
+        _metadataKey: _metadata.toYaml(),
         _versionKey: version.toString(),
       }..sortOnKey();
 
-  String toYamlString() => yamlEncode(toYaml());
+  String toYamlString(Version version) => yamlEncode(toYaml(version));
 
-  /// The version of [BuildOutput].
+  /// The version of [BuildOutputImpl].
   ///
-  /// This class is used in the protocol between the Dart and Flutter SDKs
-  /// and packages through `build.dart` invocations.
+  /// This class is used in the protocol between the Dart and Flutter SDKs and
+  /// packages through `build.dart` invocations.
   ///
   /// If we ever were to make breaking changes, it would be useful to give
   /// proper error messages rather than just fail to parse the YAML
   /// representation in the protocol.
-  static Version version = Version(1, 0, 0);
+  ///
+  /// [BuildOutput.latestVersion] is tied to [BuildConfig.latestVersion]. This
+  /// enables making the yaml serialization in `build.dart` dependent on the
+  /// version of the Dart or Flutter SDK. When there is a need to split the
+  /// versions of BuildConfig and BuildOutput, the BuildConfig should start
+  /// passing the highest supported version of BuildOutput.
+  static Version latestVersion = BuildConfigImpl.latestVersion;
 
   static const fileName = 'build_output.yaml';
 
-  /// Writes the YAML file from [outDir]/[fileName].
-  static Future<BuildOutput?> readFromFile({required Uri outDir}) async {
+  /// Reads the YAML file from [outDir]/[fileName].
+  static Future<BuildOutputImpl?> readFromFile({required Uri outDir}) async {
     final buildOutputUri = outDir.resolve(fileName);
     final buildOutputFile = File.fromUri(buildOutputUri);
     if (!await buildOutputFile.exists()) {
       return null;
     }
-    return BuildOutput.fromYamlString(await buildOutputFile.readAsString());
+    return BuildOutputImpl.fromYamlString(await buildOutputFile.readAsString());
   }
 
-  /// Writes the [toYamlString] to [outDir]/[fileName].
-  @override
-  Future<void> writeToFile({required Uri outDir}) async {
+  /// Writes the [toYamlString] to [BuildConfig.outputDirectory]/[fileName].
+  Future<void> writeToFile({required BuildConfig config}) async {
```

Comment:
Does `BuildConfig`  / `BuildOutput` really need to depend on `dart:async` and `dart:io`?

Imagine you have
```dart
main(args) async {
  await build((BuildConfig config, BuildOutput output) {
  });
}

// input config
class BuildConfig {
  final Map json;
  BuildConfig(this.json);
}

extension on BuildConfig {
  Native get native = Native(json['native']);
}

class Native {
  final Map json;
  BuildConfig(this.json);

  Abi get abi => ...;
}

// output
class BuildOutput {
  ...
  String toJson();
}
```

The only piece here that needs to do asynchronous things and I/O is `performBuild` (which doesn't have to be in the same library (or even package) as the protocol)